google/cloud/internal/big_endian.h

```diff
@@ -15,164 +15,67 @@
-                                 (((num_value)&0x00000000ff000000ull) << 8) |
-                                 (((num_value)&0x0000000000ff0000ull) << 24) |
-                                 (((num_value)&0x000000000000ff00ull) << 40) |
-                                 (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
-  }
-
-  static inline std::uint16_t byte_swap(std::uint16_t value) {
-#ifdef _MSC_VER
-    return _byteswap_ushort(value);
-#elif defined(__APPLE__)
-    return OSSwapInt16(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap16(value);
-#else
-    std::uint16_t return_value =
-        ((((num_value)&0xff00u) >> 8) | (((num_value)&0x00ffu) << 8));
 
-    return return_value;
-#endif
-  }
-
-  static inline std::uint32_t byte_swap(std::uint32_t value) {
-#ifdef _MSC_VER
-    return _byteswap_ulong(value);
-#elif defined(__APPLE__)
-    return OSSwapInt32(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap32(value);
-#else
-    std::uint32_t return_value =
-        ((((num_value)&0xff000000ul) >> 24) |
-         (((num_value)&0x00ff0000ul) >> 8) | (((num_value)&0x0000ff00ul) << 8) |
-         (((num_value)&0x000000fful) << 24));
-
-    return return_value;
-#endif
-  }
-
-  static inline std::uint64_t byte_swap(std::uint64_t value) {
-#ifdef _MSC_VER
-    return _byteswap_uint64(value);
-#elif defined(__APPLE__)
-    return OSSwapInt64(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap64(value);
-#else
-    std::uint64_t return_value = ((((num_value)&0xff00000000000000ull) >> 56) |
-                                  (((num_value)&0x00ff000000000000ull) >> 40) |
-                                  (((num_value)&0x0000ff0000000000ull) >> 24) |
-                                  (((num_value)&0x000000ff00000000ull) >> 8) |
-                                  (((num_value)&0x00000000ff000000ull) << 8) |
-                                  (((num_value)&0x0000000000ff0000ull) << 24) |
-                                  (((num_value)&0x000000000000ff00ull) << 40) |
-                                  (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
+// Encodes an 8-bit, 16-bit, 32-bit, or 64-bit signed or unsigned value as a
+// big-endian sequence of bytes. The returned string has a size matching
+// `sizeof(T)`. Example:
+//
+//   std::string s = EncodeBigEndian(std::int32_t{255});
+//   assert(s == std::string("\0\0\0\xFF", 4));
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+inline std::string EncodeBigEndian(T value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                "This code assumes an 8-bit char");
+  using unsigned_type = typename std::make_unsigned<T>::type;
+  unsigned_type const n = value;
+  unsigned_type shift = sizeof(n) * 8;
```

Comment:
Should the type of `shift` be unrelated to the type of `T` (e.g., just `size_t`)?

Reply:
Good point. Done. Well, I made the type `auto`, which should make it `size_t` due o the use of sizeof().